### PR TITLE
Implement .at property for Koalas DataFrames and Series

### PR DESCRIPTION
**Pull request by floscha**
As requested in #382, this PR implements pandas' `.at` property for Koalas DataFrames and Series.

### CODE REVIEW

File: databricks/koalas/indexing.py [hunk 1]
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
...
+        from databricks.koalas.frame import DataFrame
+
+        if self._ks is None and len(key) != 2:
+            raise TypeError("Use DataFrame.at like .at[row_index, column_name]")
+        if self._ks is not None and len(key) != 1:
+            raise TypeError("Use Series.at like .at[column_name]")
+
+        column = key[1] if len(key) > 1 else self._ks.name
+        if column is not None and column not in self._kdf._metadata.data_columns:
+            raise KeyError("'%s" % column)
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: just out of curiosity, what's `'` for?

File: databricks/koalas/indexing.py [hunk 2]
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
...
+        series = self._ks if self._ks is not None else self._kdf[column]
+
+        row = key[0]
+        sdf = (series._kdf._sdf
+               .where(F.col(self._kdf._metadata.index_columns[0]) == row)
+               .select(column))
+        if sdf.count() < 1:
+            raise KeyError("'%s" % row)
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: ditto.

File: databricks/koalas/indexing.py [hunk 3]
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
...
+
+        values = DataFrame(sdf).to_pandas().iloc[:, 0].values
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: Actually we don't need to create Koalas `DataFrame` here. Spark DataFrame has `toPandas()`.

File: databricks/koalas/indexing.py [hunk 4]
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
...
+            self._kdf = df_or_s._kdf
+            self._ks = df_or_s
+
+    def __getitem__(self, key):
+        from databricks.koalas.frame import DataFrame
+
+        if self._ks is None and len(key) != 2:
+            raise TypeError("Use DataFrame.at like .at[row_index, column_name]")
+        if self._ks is not None and len(key) != 1:
+            raise TypeError("Use Series.at like .at[column_name]")
```
[the remark this answers is not in the document]
Comment (floscha) [reply databricks/koalas/indexing.py]: True. Fixed with 8137630.

File: databricks/koalas/indexing.py [hunk 5]
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
...
+        else:
+            # If df_or_col is Column, store both the DataFrame anchored to the Column and
+            # the Column itself.
+            self._kdf = df_or_s._kdf
+            self._ks = df_or_s
+
+    def __getitem__(self, key):
+        from databricks.koalas.frame import DataFrame
+
+        if self._ks is None and len(key) != 2:
```
[the remark this answers is not in the document]
Comment (floscha) [reply databricks/koalas/indexing.py]: Sure, added with 09df062.

File: databricks/koalas/indexing.py [hunk 6]
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
...
+            raise TypeError("Use Series.at like .at[column_name]")
+
+        column = key[1] if len(key) > 1 else self._ks.name
+        if column is not None and column not in self._kdf._metadata.data_columns:
+            raise KeyError("'%s" % column)
+        series = self._ks if self._ks is not None else self._kdf[column]
+
+        row = key[0]
+        sdf = (series._kdf._sdf
+               .where(F.col(self._kdf._metadata.index_columns[0]) == row)
```
[the remark this answers is not in the document]
Comment (floscha) [reply databricks/koalas/indexing.py]: I've added a check now that raises an exception if the index level doesn't equal 1 with a61f163. I'll have to look closer into how `.at` behaves for multilevel indices which generally seems to be rather uncommon. Also, is a Koalas DataFrame without an index even valid? Like, wouldn't it for example crash when calling `to_pandas()` on it?

File: databricks/koalas/indexing.py [hunk 7]
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
...
+               .select(column))
+        if sdf.count() < 1:
```
[the remark this answers is not in the document]
Comment (floscha) [reply databricks/koalas/indexing.py]: Good point! I've implemented your suggestion in 2dc8fa5.

File: databricks/koalas/indexing.py (hunk 3 above)
Comment (floscha) [reply databricks/koalas/indexing.py]: You're right. Fixed this with 2dc8fa5.

File: databricks/koalas/indexing.py (hunk 1 above)
Comment (floscha) [reply databricks/koalas/indexing.py]: This is just to be consistent with the way pandas uses Python's `KeyError`, putting the missing key between `'`s. However, I've just realized I've missed the closing `'` and the `'`s are not even necessary to start with 🙈 So I've fixed this with 2dc8fa5.

File: databricks/koalas/indexing.py (hunk 2 above)
Comment (floscha) [reply databricks/koalas/indexing.py]: See comment above.

File: databricks/koalas/indexing.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ueshin) [reply databricks/koalas/indexing.py]: As for multi- or no index, currently `.loc` raises an exception for multi- or no index. Maybe we can follow the behavior for now, but please make sure to check it and raise the exception.
And yes, no index in Koalas DataFrame is valid, but I'm not sure whether we should crash in that case. E.g., `sdf.to_koalas()` doesn't have index, but should we explicitly `set_index` prior to `to_pandas()` in that case? This is the future work anyway. Let's file an issue to discuss it.
Thanks!

**Comment by floscha** [comment]
Thanks for your thorough review @ueshin!

I've addressed all of your comments and also added a warning note for matching a lot of rows with f14d136.

File: databricks/koalas/indexing.py [hunk 8]
```diff
@@ -82,6 +82,80 @@ def _unfold(key, kseries):
...
+            # the Column itself.
+            self._kdf = df_or_s._kdf
+            self._ks = df_or_s
+
+    def __getitem__(self, key):
+        from databricks.koalas.frame import DataFrame
+
+        if self._ks is None and (not isinstance(key, tuple) or len(key) != 2):
+            raise TypeError("Use DataFrame.at like .at[row_index, column_name]")
+        if self._ks is not None and len(key) != 1:
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: Need to check key is not str?

File: databricks/koalas/indexing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ueshin) [reply databricks/koalas/indexing.py]: Could you add tests for such a case?

File: databricks/koalas/indexing.py (hunk 8 above)
Comment (ueshin) [reply databricks/koalas/indexing.py]: I mean what if `kdf.a.at['abc']`, for example?

File: databricks/koalas/indexing.py [hunk 9]
```diff
@@ -82,6 +82,80 @@ def _unfold(key, kseries):
     return rows_sel, cols_sel
 
 
+class AtIndexer(object):
+    """
+    Access a single value for a row/column label pair.
+    Similar to ``loc``, in that both provide label-based lookups. Use ``at`` if you only need to
+    get a single value in a DataFrame or Series.
```
Comment (thunterdb) [inline databricks/koalas/indexing.py]: you should also mention the behavior when multiple values are matched (returning an array)

File: databricks/koalas/indexing.py [hunk 10]
```diff
@@ -82,6 +82,80 @@ def _unfold(key, kseries):
...
+    ...                    index=[4, 5, 6], columns=['A', 'B', 'C'])
+    >>> kdf
+        A   B   C
+    4   0   2   3
+    5   0   4   1
+    6  10  20  30
+
+    Get value at specified row/column pair
+
+    >>> kdf.at[4, 'B']
```
Comment (thunterdb) [inline databricks/koalas/indexing.py]: can you also put an example with multiple values matching the same index, to show that it returns an array?

File: databricks/koalas/indexing.py (hunk 9 above)
Comment (floscha) [reply databricks/koalas/indexing.py]: I've added it to the docs with 0dc9e08.

File: databricks/koalas/indexing.py (hunk 10 above)
Comment (floscha) [reply databricks/koalas/indexing.py]: Sure, done with 83a77d0.

File: databricks/koalas/indexing.py (hunk 8 above)
Comment (floscha) [reply databricks/koalas/indexing.py]: Good point! I've added the check and an additional unit test (that used to fail previously) with 
415e78a.

File: databricks/koalas/indexing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (floscha) [reply databricks/koalas/indexing.py]: Sure, done with fb3587d.

File: databricks/koalas/indexing.py [hunk 11]
```diff
@@ -82,6 +82,84 @@ def _unfold(key, kseries):
...
+    5  10  20  30
+
+    Get value at specified row/column pair
+
+    >>> kdf.at[4, 'B']
+    2
+
+    Get array if an index occurs multiple times
+
+    >>> kdf.to_pandas().at[5, 'B']
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: `kdf.at[5, 'B']` ?

File: databricks/koalas/tests/test_indexing.py [hunk 12]
```diff
@@ -126,6 +126,45 @@ def pdf(self):
...
+        self.assertEqual(kdf.at[3, 'b'], 6)
+        self.assertEqual(kdf.at[3, 'b'], pdf.at[3, 'b'])
+        np.testing.assert_array_equal(kdf.at[9, 'b'], np.array([0, 0, 0]))
+        np.testing.assert_array_equal(kdf.at[9, 'b'], pdf.at[9, 'b'])
+
+        # Assert .at for Series
+        self.assertEqual(test_series.at['b'], 6)
+        self.assertEqual(test_series.at['b'], pdf.loc[3].at['b'])
+
+        #
```
Comment (ueshin) [inline databricks/koalas/tests/test_indexing.py]: forgot to add a comment?

File: databricks/koalas/indexing.py (hunk 11 above)
Comment (floscha) [reply databricks/koalas/indexing.py]: You're right. Fixed with ebea010.

File: databricks/koalas/tests/test_indexing.py (hunk 12 above)
Comment (floscha) [reply databricks/koalas/tests/test_indexing.py]: Correct. Guess I was kind of tired yesterday when committing this... Now fixed with ebea010 though.

**Comment by ueshin** [comment]
Thanks! merging.

**Pull request merged**